### PR TITLE
Rename method on permission service

### PR DESCRIPTION
**Pull request by hmlanigan**
Follow on to #16894.

Repurposing the Update method of the permission service to be UpdateMaybeCreatePermission. All targets for permissions have similar logic to grant or revoke permissions and maybe add the user: ModifyCloudAccess, ChangeControllerAccess, ModifyOfferAccess and ModifyModelAccess. This will allow the business logic of these methods to be made target independent and pushed out of the apiserver facades and into the domain.
 
## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Unit tests

## Links

Need discovered during JUJU-5476

### CODE REVIEW

File: domain/permission/service/service.go [hunk 1]
```diff
@@ -17,15 +18,17 @@ type State interface {
...
 	DeletePermission(ctx context.Context, subject string, target permission.ID) error
 
-	// UpdatePermission updates the subject's (user) access to the given
-	// target to the given access.
-	UpdatePermission(ctx context.Context, subject string, target permission.ID, access permission.Access) error
+	// UpdateMaybeCreatePermission updates the permission on the target for the given
+	// subject (user). The api user must have Admin permission on the target. If a
+	// subject does not exist, it is created using the subject and api user. Access
+	// can be granted or revoked.
+	UpdateMaybeCreatePermission(ctx context.Context, args state.UpdateMaybeCreatePermissionArgs) error
```
Comment (SimonRichardson) [inline domain/permission/service/service.go:31]: Upsert is the common name for this action.

**Review by SimonRichardson — COMMENTED** [review]
I think we should just call this upsert.

File: domain/permission/service/service.go (hunk 1 above)
Comment (hmlanigan) [reply domain/permission/service/service.go]: done

File: domain/permission/service/service.go [hunk 2]
```diff
@@ -17,15 +18,17 @@ type State interface {
 	// CreatePermission gives the user access per the provided spec.
 	// It requires the user/target combination has not already been
 	// created.
-	CreatePermission(ctx context.Context, spec UserAccessSpec) (permission.UserAccess, error)
+	CreatePermission(ctx context.Context, spec state.UserAccessSpec) (permission.UserAccess, error)
```
Comment (SimonRichardson) [inline domain/permission/service/service.go:21]: FYI: we have a rule that service doesn't know about the state package and the service package doesn't know about state. So in this case you could have a file called userspec.go in the root of the `domain/permission` package, that way they don't depend on each other. Alternatively, you could move it to core if it has more use outside of the domain.

File: domain/permission/service/service.go [hunk 3]
```diff
@@ -179,16 +186,56 @@ func (s *Service) ReadAllAccessTypeForUser(ctx context.Context, subject string,
...
-// is empty, or the target is not valid. Any errors from the state layer
-// are passed through.
-func (s *Service) UpdatePermission(ctx context.Context, subject string, target permission.ID, access permission.Access) error {
-	if subject == "" {
+type AccessChange string
+
+const (
+	Grant  AccessChange = "grant"
+	Revoke AccessChange = "revoke"
+)
```
Comment (SimonRichardson) [inline domain/permission/service/service.go:194]: These feel like they should live in core.

**Review by SimonRichardson — COMMENTED** [review]
Last bit, otherwise all good.

File: domain/permission/service/service.go (hunk 3 above)
Comment (hmlanigan) [reply domain/permission/service/service.go:194]: agreed

File: domain/permission/service/service.go (hunk 2 above)
Comment (hmlanigan) [reply domain/permission/service/service.go:21]: done

**Review by SimonRichardson — APPROVED** [review]
Excellent.

**Pull request merged**